Nb port content type enrichment from capi

NathanielBennett opened the pull request:
@mchv This is the change you were discussing with @alexduf . It ports the logic in a (reverted) MAPI PR: https://github.com/guardian/mobile-apps-api/pull/1021 * to implement the Garnett contentType logic on the api-client. 

*This is reverted because 'contentType' clashes with a attribute on a client-side object which caused the app to crash. We'll reimplement using a new attribute name. 

Why does updating the changelog break the build?

@alexduf

mchv commented on src/main/scala/com.gu.contentapi.client/utils/ContentType.scala:
```diff
@@ -0,0 +1,20 @@
+package com.gu.contentapi.client.utils
+
+sealed trait ContentType
```
Should be called `DesignType` as we already have `ContentType` in the model

mchv commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 11.43 
+
+* Adds an implicit contentType field to Content model
```
should be named `designType`

mchv commented on src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:
```diff
@@ -15,4 +14,46 @@ object CapiModelEnrichment {
+
+    def isComment = tagExistsWithId("tone/comment") || tagExistsWithId("tone/letters")
+
+    def contentType: ContentType = {
```
`designType: DesignType`

mchv commented on src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:
```diff
@@ -15,4 +14,46 @@ object CapiModelEnrichment {
+        filter( Some(Live) ) { c => c.tagExistsWithId("tone/minutebyminute")}
+      )
+
+      def findContentType(contentFilters: List[DesignTypeFilter], c: Content) : ContentType = contentFilters match {
```
`findDesignType(contentFilters: List[DesignTypeFilter], c: Content) : DesignType`

mchv commented on src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:
```diff
@@ -15,4 +14,46 @@ object CapiModelEnrichment {
 
+  implicit class RichContent(val content: Content) extends AnyVal {
+
+    private def defaultContentType = Article
```
should be a `val` and  name `defaultDesignType`

mchv commented:
Thanks @NathanielBennett!

mchv commented:
Apart from the name of the new field, this looks good! May you update the code and I will have another quick review before merging.

NathanielBennett commented:
@mchv Take a look now. I've changed the name of the new field. Have also made the helper methods I implemented in the implicit class private to the designType function so as not to pollute the `Content` interface. It might be a little, tiny-bit less readable

mchv commented on src/test/scala/com.gu.contentapi.client/model/utils/CapiModelEnrichmentTest.scala:
```diff
@@ -0,0 +1,161 @@
+
+class CapiModelEnrichmentTest extends FlatSpec with MockitoSugar with Matchers {
+
+  it should  "have a contentType of 'Media' when tag type/video is present" in {
```
`have a designType`

NathanielBennett replied:
Damnit

NathanielBennett replied:
Sorry, that was uncalled for. Fixed now :(

tomrf1 commented on src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:
```diff
@@ -15,4 +14,46 @@ object CapiModelEnrichment {
+        c => if(contentFilter(c)) result else None
+      }
+
+      val fiters: List[DesignTypeFilter] = List(
```
typo on `fiters` :)

NathanielBennett replied:
thanks!

tomrf1 commented on src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:
```diff
@@ -15,4 +14,46 @@ object CapiModelEnrichment {
+        filter( Some(Live) ) ( tagExistsWithId("tone/minutebyminute") )
+      )
+
+      def findDesignType(contentFilters: List[DesignTypeFilter], c: Content) : DesignType = contentFilters match {
```
If you change `isMedia` etc to vals then you could do something a tiny bit shorter, e.g.:
```
val isMedia: Content => Boolean = c => tagExistsWithId("type/audio")(c) || tagExistsWithId("type/video")(c) || tagExistsWithId("type/gallery")(c)
...
val predicates: Map[Content => Boolean, DesignType] = Map(
  isImmersive -> Immersive,
  isMedia -> Media,
  isReview -> Review,
  tagExistsWithId("tone/analysis") -> Analysis,
  isComment -> Comment,
  tagExistsWithId("tone/features") -> Feature,
  tagExistsWithId("tone/minutebyminute") -> Live
)

val result = predicates.collectFirst { case (predicate, design) if predicate(content) => design }
result.getOrElse(defaultDesignType)
```

NathanielBennett replied:
That's cool .. :)

NathanielBennett replied:
Except i can't get it to compile . Might give you a shout tomorrow mornign

tomrf1 replied:
Oops, didn't realise the order is important. The following works locally for me, what do you think? Happy to chat tomorrow!
```
def designType: DesignType = {

  val defaultDesignType = Article

  val isImmersive: Content => Boolean = c => c.fields.flatMap(_.displayHint).exists(_.equals("immersive"))
  def tagExistsWithId(tagId: String): Content => Boolean = c => c.tags.exists(tag => tag.id == tagId)
  val isMedia: Content => Boolean = c => tagExistsWithId("type/audio")(c) || tagExistsWithId("type/video")(c) || tagExistsWithId("type/gallery")(c)
  val isReview: Content => Boolean = c => tagExistsWithId("tone/reviews")(c) || tagExistsWithId("tone/livereview")(c) || tagExistsWithId("tone/albumreview")(c)
  val isComment: Content => Boolean = c => tagExistsWithId("tone/comment")(c) || tagExistsWithId("tone/letters")(c)

  val predicates: List[(Content => Boolean, DesignType)] = List(
    isImmersive -> Immersive,
    isMedia -> Media,
    isReview -> Review,
    tagExistsWithId("tone/analysis") -> Analysis,
    isComment -> Comment,
    tagExistsWithId("tone/features") -> Feature,
    tagExistsWithId("tone/minutebyminute") -> Live
  )

  val result = predicates.collectFirst { case (predicate, design) if predicate(content) => design }
  result.getOrElse(defaultDesignType)
}
```

NathanielBennett commented:
Hey can you hold back on merging this??

mchv commented:
I think Map will not worked becasue of the order, you will need SortedMap



Le jeu. 23 nov. 2017 à 17:46, Nathaniel Bennett <notifications@github.com>
a écrit :

> *@NathanielBennett* commented on this pull request.
> ------------------------------
>
> In src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
> <https://github.com/guardian/content-api-scala-client/pull/239#discussion_r152853821>
> :
>
> > +
> +      def filter(result: Option[DesignType])(contentFilter: Content => Boolean) : DesignTypeFilter = {
> +        c => if(contentFilter(c)) result else None
> +      }
> +
> +      val filters: List[DesignTypeFilter] = List(
> +        filter( Some(Immersive) ) ( isImmersive ),
> +        filter( Some(Media) ) ( isMedia ),
> +        filter( Some(Review) ) ( isReview ),
> +        filter( Some(Analysis) ) ( tagExistsWithId("tone/analysis") ),
> +        filter( Some(Comment) ) ( isComment ),
> +        filter( Some(Feature) ) ( tagExistsWithId("tone/features") ),
> +        filter( Some(Live) ) ( tagExistsWithId("tone/minutebyminute") )
> +      )
> +
> +      def findDesignType(contentFilters: List[DesignTypeFilter], c: Content) : DesignType = contentFilters match {
>
> Except i can't get it to compile . Might give you a shout tomorrow mornign
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/guardian/content-api-scala-client/pull/239#discussion_r152853821>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAlirXnsTWDRHu8BOXbhmkBs0gk9dqT6ks5s5a-LgaJpZM4Qo3-v>
> .
>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

NathanielBennett commented:
Nah it's not the map . I didn't get that far. Changing the `defs` to `vals` causes a wrong forward reference compilation error.

NathanielBennett commented:
@tomrf1 That works good! Thanks! 
@mchv Are we good to go now?

tomrf1 commented:
👍 💯

NathanielBennett commented:
Nice! Do you guys release it or do I?

tomrf1 commented:
@NathanielBennett I'll release now

NathanielBennett commented:
Thanks @tomrf1 !
